### DISCUSSION TITLE
How can a python package safely support both of pydantic v1 and v2?

### DISCUSSION THREAD

**Discussion by yu-iskw** [pydantic/pydantic#8203] [Question]
Some other contributors and I created and maintain a python package which depends on pydantic1 today. As a package developer with pydantic, I want to know the way to safely support pydantic 1 and 2, since some users of the package use pydantic 1 and the others use pydantic v2. And I want to consider situations that some users with pydantic v1 can't immediately migrate to pydantic v2 because of their dependencies. There would be some potential solutions to solve the issue. For instance, we can manage different versinons of the packaged based on the pydantic version as well as we can provide another python package with pydantic 2. But, ideally, a version of the package can support both of pydantic v1 and v2 at the same time. Can you give me advice to safely support both?

- https://github.com/yu-iskw/dbt-artifacts-parser

I found a similar discussion. But, it's slightly different.
https://github.com/pydantic/pydantic/discussions/4649

**Comment by lvzhengcq** [comment]
Supporting multiple versions of a library, such as Pydantic in this case, can be challenging due to potential breaking changes between versions. However, you can use conditional imports and version checks to manage compatibility. Here's a general approach:

Check Pydantic Version:
Use conditional imports to import the appropriate parts of Pydantic based on the version. You can check the Pydantic version dynamically within your package.

python
import pydantic

if pydantic.__version__.startswith('1.'):
    from pydantic_v1_module import SomeClass
elif pydantic.__version__.startswith('2.'):
    from pydantic_v2_module import SomeClass
else:
    raise ImportError("Unsupported Pydantic version")

**Reply by yu-iskw** [reply]
@lvzhengcq Thank you for answering my question. Dynamically switching import statements based on the environment of a user would be a possible approach. I understand we need something like that, no matter what approach we take.

**Comment by yu-iskw** [comment]
We don't have any smart way to support multiple pydantic versions with perfect compatibilities as of this writing. Let me close the discussion.
